### PR TITLE
Fixes a major and a minor bug in the dockerized gamma

### LINKED ISSUE #96: Dockerized gamma cannot handle parallel requests reliable due to logging file system (wrapperList.json)
There is a wrapperList.json file in the root workspaces folder, that lists the projects and the workspaces. This log file is not handled atomically with the corresponding file system operations (e.g. folder creation and deletion).

If the dockerized gamma server receives two parallel requests, then this file might be overriden concurrently. It means, if one of the requests adds/removes a workspace (or a project in a workspace), then the changes made by one of the requests might get lost because the other reqest overrides the file and deletes these changes.

As the code base heavily relies on this file to check which projects and workspaces exist on the server, it is crucial to fix this concurrency bug.

### LINKED ISSUE #98: Stop operation throws NullPointerException in dockerized Gamma
```java
Dec 07, 2021 3:28:34 PM hu.bme.mit.gamma.headless.server.OpenApiWebServer lambda$start$5
INFO: Operation "stopOperation" has started.
Dec 07, 2021 3:28:34 PM io.vertx.ext.web.RoutingContext
SEVERE: Unhandled exception in router
java.lang.NullPointerException
        at hu.bme.mit.gamma.headless.server.OpenApiWebServer.sendErrorResponse(OpenApiWebServer.java:635)
        at hu.bme.mit.gamma.headless.server.OpenApiWebServer.lambda$start$5(OpenApiWebServer.java:285)
        at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1127)
        at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:114)
        at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:133)
        at io.vertx.ext.web.validation.impl.ValidationHandlerImpl.lambda$handle$5(ValidationHandlerImpl.java:168)
        at io.vertx.core.impl.future.FutureBase.emit(FutureBase.java:96)
        at io.vertx.core.impl.future.SucceededFuture.onComplete(SucceededFuture.java:75)
        at io.vertx.ext.web.validation.impl.ValidationHandlerImpl.handle(ValidationHandlerImpl.java:159)
        at io.vertx.ext.web.validation.impl.ValidationHandlerImpl.handle(ValidationHandlerImpl.java:18)
        at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1127)
        at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:151)
        at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:133)
        at io.vertx.ext.web.handler.impl.BodyHandlerImpl$BHandler.doEnd(BodyHandlerImpl.java:313)
        at io.vertx.ext.web.handler.impl.BodyHandlerImpl$BHandler.end(BodyHandlerImpl.java:290)
        at io.vertx.ext.web.handler.impl.BodyHandlerImpl.lambda$handle$0(BodyHandlerImpl.java:86)
        at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:96)
        at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:59)
        at io.vertx.core.http.impl.HttpEventHandler.handleEnd(HttpEventHandler.java:76)
        at io.vertx.core.http.impl.Http1xServerRequest.onEnd(Http1xServerRequest.java:556)
        at io.vertx.core.http.impl.Http1xServerRequest.handleEnd(Http1xServerRequest.java:542)
        at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:70)
        at io.vertx.core.impl.DuplicatedContext.execute(DuplicatedContext.java:162)
        at io.vertx.core.http.impl.Http1xServerConnection.onEnd(Http1xServerConnection.java:199)
        at io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Http1xServerConnection.java:142)
        at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:155)
        at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:154)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
        at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:99)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:719)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:655)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:581)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

### PR DESCRIPTION
**Pull request by benedekh**
Fixes #96.
Fixes #98.

Major bug: #96 
Minor bug: #98

### CODE REVIEW

File: plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/service/Validator.java [hunk 1]
```diff
@@ -69,6 +64,9 @@ public static boolean checkIfProjectHasRunIntoError(String workspace, String pro
 	public static boolean checkIfProjectIsUnderLoad(String workspace, String projectName) throws IOException {
 		File jsonFile = new File(FileHandlerUtil.getProperty(DIRECTORY_OF_WORKSPACES_PROPERTY_NAME) + workspace
 				+ File.separator + projectName + File.separator + PROJECT_DESCRIPTOR_JSON);
+		if(!jsonFile.exists()) {
```
Comment (grbeni) [inline plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/service/Validator.java:67]: Add space after `if`

Comment (benedekh) [reply plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/service/Validator.java]: fixed in ebe98c2

File: plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/OpenApiWebServer.java [hunk 2]
```diff
@@ -633,27 +637,29 @@ private void sendErrorResponse(RoutingContext routingContext, ErrorHandlerPOJO e
...
 	private static void listWorkspacesAndProjects() throws IOException {
-		if (FileHandlerUtil.getWrapperListFromJson() != null) {
-			List<WorkspaceProjectWrapper> yourList = FileHandlerUtil.getWrapperListFromJson();
+		Map<String, Set<String>> projectsByWorkspaces = FileHandlerUtil.getProjectsByWorkspaces();
+		if (projectsByWorkspaces.isEmpty()) {
+			logger.log(Level.INFO, "No workspaces are present.");
+			return;
+		}
+		
+		for(Map.Entry<String, Set<String>> projectsByWorkspace: projectsByWorkspaces.entrySet()) {
```
Comment (grbeni) [inline plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/OpenApiWebServer.java:646]: Add space after `for`

File: plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsDirectoryFilter.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+import java.io.FileFilter;
+
+public class IsDirectoryFilter implements FileFilter {
+	
+	@Override
+	public boolean accept(File file) {
+		return file.isDirectory();
+	}
+	
+	public static IsDirectoryFilter Create() {
```
Comment (grbeni) [inline plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsDirectoryFilter.java:13]: Why is the first character of the method name capitalized?

File: plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsDirectoryFilter.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package hu.bme.mit.gamma.headless.server.util;
+
+import java.io.File;
+import java.io.FileFilter;
+
+public class IsDirectoryFilter implements FileFilter {
```
Comment (grbeni) [inline plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsDirectoryFilter.java:6]: Wouldn't `DirectoryFilter` be a better name?

File: plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsEclipseProjectFilter.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+import java.io.File;
+import java.io.FileFilter;
+import java.io.FilenameFilter;
+
+public class IsEclipseProjectFilter implements FileFilter {
+
+	@Override
+	public boolean accept(File file) {
+		if(file.isDirectory()) {
```
Comment (grbeni) [inline plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsEclipseProjectFilter.java:11]: Add space

File: plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsEclipseProjectFilter.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
+package hu.bme.mit.gamma.headless.server.util;
+
+import java.io.File;
+import java.io.FileFilter;
+import java.io.FilenameFilter;
+
+public class IsEclipseProjectFilter implements FileFilter {
```
Comment (grbeni) [inline plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsEclipseProjectFilter.java:7]: Wouldn't `EclipseProjectFilter` be a better name?

File: plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsDirectoryFilter.java (hunk 3 above)
Comment (grbeni) [reply plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsDirectoryFilter.java:13]: Also, couldn't this class be a singleton?

File: plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsEclipseProjectFilter.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+				}
+				
+			});
+			return eclipseProjectFiles.length > 0;
+		}
+		return false;
+	}
+	
+	public static IsEclipseProjectFilter Create() {
+		return new IsEclipseProjectFilter();
```
Comment (grbeni) [inline plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsEclipseProjectFilter.java:25]: See `DirectoryFilter`

**Review by grbeni — CHANGES_REQUESTED** [review]

File: plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsEclipseProjectFilter.java (hunk 5 above)
Comment (benedekh) [reply plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsEclipseProjectFilter.java]: fixed in 9ce2b0c

File: plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/OpenApiWebServer.java (hunk 2 above)
Comment (benedekh) [reply plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/OpenApiWebServer.java]: fixed in 9ce2b0c

File: plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsEclipseProjectFilter.java (hunk 7 above)
Comment (benedekh) [reply plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsEclipseProjectFilter.java]: fixed in 82e3b4f

File: plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsEclipseProjectFilter.java (hunk 6 above)
Comment (benedekh) [reply plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsEclipseProjectFilter.java]: fixed in 82e3b4f

File: plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsDirectoryFilter.java (hunk 4 above)
Comment (benedekh) [reply plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsDirectoryFilter.java]: fixed in 82e3b4f

File: plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsDirectoryFilter.java (hunk 3 above)
Comment (benedekh) [reply plugins/headless/hu.bme.mit.gamma.headless.server/src/hu/bme/mit/gamma/headless/server/util/IsDirectoryFilter.java]: fixed in 82e3b4f

**Review by grbeni — APPROVED** [review]

**Pull request merged**